config/config.go

```diff
@@ -0,0 +1,81 @@
+package config
+
+import (
+	"fmt"
+	"os"
+)
+
+type DBConfig struct {
+	Database string
+	User     string
+	Password string
+	Host     string
+}
+
+type OauthConfig struct {
+	Secret      string
+	Client      string
+	RedirectUrl string
+	State       string
+}
+
+type Server struct {
+	SessionCookieName string
+	CookieSecret      string
+}
+
+type Cache struct {
+	Address  string
+	Password string
+}
+
+type Config struct {
+	DB     DBConfig
+	Oauth  OauthConfig
+	Server Server
+	Cache  Cache
+}
+
+// New returns a new Config struct
+func New() *Config {
```

Comment:
I am struggling with this one a bit in terms of the config specifically.If there are use cases where I do not want the app to continue (ex: database connection missing) as it would lead to an eventual crash of the application anywhere, wouldn't it be acceptable to panic there? 

I do get some of the other uses of panic not being appropriate though and will clean those up. 

The config at a high level like this is causing a headache because it is passed so many levels deep. I am reworking it so that each package can get its own config values as other packages really do not need to have access to another's config values. This is making it a little cleaner now that database is responsible for its errors while configuring.

Reply:
Now that I have the DI bit done, I think I can actually do this easier than I thought if I pass the errors back from the New* methods from each creation and then I can use them in main. let me give this a try and see how it looks.

Reply:
Typically what I've seen is two levels of validation. One at the config level where it's making sure everything is set, and another at the service/component level where, for example, it's checking that it can actually connect to the database.

In both cases, I'd bubble up an error to your main function and do something like
```go
db, err := NewDB(cfg.DB)
if err != nil {
    log.Fatalf("Failed to initialize database: %w", err)
}
```
which will log the error and exit the process with a status code of 1 (error).